Step2

nfreedom9 opened the pull request:
2단계 - 문자열 계산기

nfreedom9 committed:
패키지명 및 파일 위치 변경

nfreedom9 committed:
StringParser, Operators 구현 및 테스트 코드 작성

nfreedom9 committed:
계산기 구현 및 테스트 코드

nfreedom9 committed:
클래스명 변경

laco-dev commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,17 @@
+package step2
+
+class Operator {
+    companion object {
+        val list = listOf<String>("+", "-", "*", "/")
+        val plus = { a: Double, b: Double -> a + b }
+        val subtract = { a: Double, b: Double -> a - b }
+        val multiply = { a: Double, b: Double -> a * b }
+        const val division0ErrorMessage = "0 으로 나눌 수 없습니다."
+        val division = { a: Double, b: Double ->
+            if (b == 0.0) {
+                throw IllegalArgumentException(division0ErrorMessage)
+            }
+            a / b
+        }
+    }
```
Operator 연산을 **변수**가 아닌 **객체**가 역할을 수행하도록 만들어 보면 어떨까요??

enum class를 활용해 연산자들의 종류를 정의 해 볼 수 있을 것 같아요.
```kotlin
enum class Operator {
    PLUS(...),
    SUBRACT,
    ...
}

val opeartor = Operator.PLUS
operator.operate(1, 2) // // 1 + 2
```

코틀린에서 enum class를 만들고 구현하신 변수를 프로퍼티로 가질 수 있겟죠?
https://kotlinlang.org/docs/enum-classes.html#implementing-interfaces-in-enum-classes

nfreedom9 replied:
이넘 클래스에 대해 공부하고 적용하도록 하겠습니다.

laco-dev commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,26 @@
+package step2
+
+class StringParser {
```
정적 함수만 제공하기 때문에 class가 아닌 object로 만들어 볼까요?
https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview

laco-dev commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,26 @@
+        const val emptyMessage = "입력값이 없습니다."
+        const val errorMessage = "입력값이 잘못되었습니다."
+        fun parse(string: String): List<String> {
+            if (string.isBlank()) {
+                throw IllegalArgumentException(emptyMessage)
+            }
```
도메인 로직에 따르면 비어있는 경우 오류를 발생할 수 있지만,
항상 오류를 발생시키는게 꼭 좋은 해결방법은 아닐 수 있는데요!

반환형이 리스트 타입이니 이 경우 emptyList()를 반환하도록 만들어 보면 어떨까요?

laco-dev commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,26 @@
+            if (string.isBlank()) {
+                throw IllegalArgumentException(emptyMessage)
+            }
+            val res = string.split("").joinToString("") {
+                if (it == "" || it == " ") {
+                    ""
+                } else if (Operator.list.contains(it)) {
+                    " $it "
+                } else if (it.matches("[\\d|.]".toRegex())) {
+                    it
+                } else {
+                    throw IllegalArgumentException("$errorMessage [$it]")
+                }
+            }.split(" ")
+
+            return res
```
StringParser의 역할은 입력 값으로 들어온 문자들을 띄어쓰기를 계산해서 올바른 형태의 포맷으로 만들어 주고 있는데요.

이 과정에서 Operator인지 알 필요가 있을까요??
이미 적절한 연산자가 아닌 경우 Operator에서 오류 체크가 이루어 지고 있어서 StringParser가 연산자에 강하게 의존성이 걸리고 있어요.

laco-dev replied:
간단한 로직이 생각보다 복잡하게 작성이 된 것 같아요.
입력받은 문자열의 문자를 하나씩 순회하면서 String을 만들어 줄 수 있습니다.
보다 효율적인 계산이 있을 수도 있으니 위 구조를 개선하는데 목적을 두고 고민 해 주세요.

```kotlin
val builder = StringBuilder()
string.forEach {
    when {
        it.isWhitespace() -> return@forEach
        it.isDigit() -> builder.append(it)
        else -> builder.append(" $it ")
    }
}
return builder.split(" ")
```

nfreedom9 replied:
StringParser 의 역할과 애초에 요구된 계산기 기능에 대해 제가 좀 과도하게 생각한 것 같습니다.
 
제시된 요구사항에는 입력값의 포맷에 대한 설명이 없었기 때문에,
연산자와 숫자 사이의 공백에 대하여 자유로울 수 있다고 생각했고, 
(공백이 없어도 되고, 2개 이상이어도 되는 조건)
심지어 숫자들 사이에도 공백이 가능하다고 생각했습니다.
"1 0+2   .5/5" => "10 + 2.5 / 5"
 
또한 나눗셈 연산 결과로 소수점이 나올 수 있기 때문에 입력값 역시 가능하다고 생각했습니다.
(제안해주신 것에는 소수점은 포함되지 않은 것 같습니다.)
 
때문에 입력값을 공백 기준으로 자르기 전,
기존의 공백은 모두 제거하고,
연산기호 앞뒤로만 공백을 주도록 문자열을 재정의하는 과정을 구현하게 되었습니다.
 
결국 모든 문자열을 한글자 한글자 순회하며 그것이 공백인지, 연산기호인지, 숫자인지 평가하며 처리하는 과정을 거치다보니,
자연스레 이곳에서 입력값에 대한 예외처리를 구현하게 되었습니다.

nfreedom9 replied:
역할과 책임에 대해 고민하여,
이곳에서는 문자열을 리스트로 전환해주는 역할만을 구현하도록 하고,
전환된 리스트로 실제 계산을 구현하는 Calculator 에서 예외처리를 하도록 수정하겠습니다.

laco-dev commented on src/test/kotlin/step2/StringParserTest.kt:
```diff
@@ -0,0 +1,45 @@
+    @Test
+    fun `string parse`() {
+        val stringList = listOf<String>(
+            "1.1 + 12 + 3 / 4",
+            "1.1+12+3/4",
+            " 1.1         + 1 2 + 3 / 4 ",
+            " 1 . 1         + 1 2 + 3 / 4 ",
+        )
```
여러개의 테스트 환경을 고려한 내용이라 마찬가지로 CsvSource나 ValueSource 형태로 만들어 보면 어떨까요?

nfreedom9 replied:
@ParameterizedTest 이전에 작성했던 부분인데,
@ParameterizedTest 와 일종의 비교군으로 남겨둔 것으로 이해해주시면 감사하겠습니다. ^_^;;

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+            var result = head.toDouble()
+            for (i in 1 until list.size step 2) {
+                val number = list[i + 1].toDouble()
+                val operator: (Double, Double) -> Double = when (list[i]) {
+                    "+" -> Operator.plus
+                    "-" -> Operator.subtract
+                    "*" -> Operator.multiply
+                    "/" -> Operator.division
+                    else -> throw IllegalAccessError("지원하지 않는 연산자")
+                }
```
Calculator가 연산자를 가져오는 역할과 계산하는 역할이 모두 가지고 있어요.
입력 연산자(symbol)를 통해 적절한 연산자(Operator)를 찾아오는건 연산자에게 넘겨주면 어떨까요?

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+    companion object {
+        fun calculate(string: String): Double {
+            val list = StringParser.parse(string)
+            val (head) = list
```
코틀린의 분할 선언을 이용해 첫 번째 아이템 항목을 가져온 것 같아요.
그런데 
```kotlin
val (head) = list
``` 
라는 코드만 보면 뭐지??? 라고 머릿속에 물음표가 떠있는데요.. (실제로 무엇인지 30초정도 생각함)
명시적으로 첫 번째 항목이라는 것을 표현하면 어떨까요??
```kotlin
val head = list.first()
```

laco-dev commented on src/main/kotlin/step2/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+package step2
+
+class Calculator {
```
Calculator도 object로 만들어 볼 수 있을 것 같아요

laco-dev reviewed (changes requested):
2단계 미션 수고하셨습니다.

코틀린스럽게 작성 해 보려고 노력하신 부분이 보입니다. 🤔 
코틀린스럽게 만큼 중요한건 클린스럽게.. 라고 생각합니다?

관련해서 몇 가지 구조적인 부분과 개선이 필요한 사항을 피드백 남겼습니다.

laco-dev commented:
커밋 관련해서....

1. 1단계 진행하신 이후 해당 브랜치를 리베이스 해서 다음 단계를 진행 해 주시면 됩니다.
1단계에 진행한 커밋이 2단계 PR에 그대로 남아있어요
https://github.com/next-step/kotlin-racingcar/pull/551/commits/4f37a031c2ceb4a023898e49722aa76483fd67eb

2. 커밋은 의미있는 기능 단위로 만들어 져야 합니다. 작업 중에 저장 목적으로 WIP 커밋을 남겼다면 반드시 리셋이나 리베이스 해서 정리한 뒤 진행 해 주세요.
https://github.com/next-step/kotlin-racingcar/pull/551/commits/3929476265a40c3ba79b0229da1dc88f66ce6abe

nfreedom9 committed:
리뷰 반영

- StringParser
  - class 를 object 로 변경
  - 기존의 parse 함수를 2개의 함수로 분리하고 throw 제거
  - toDouble 함수 추가.
- Operator
  - class 를 enum class 로 변경
- Calculator
  - class 를 object 로 변경
  - StringParser 에서 예외처리 하던 것을 calculate 함수에서 처리하도록 변경

nfreedom9 committed:
Calculator 에서 연산기호에 따른 연산자 선택 로직을 Operator 메소드로 변경

nfreedom9 commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,24 @@
+
+    companion object {
+        private const val symbolErrorMessage = "지원하지 않는 연산자"
+        const val division0ErrorMessage = "0 으로 나눌 수 없습니다."
```
division0ErrorMessage 변수를 private 으로 만들었다가,
에러메시지를 확인을 위한 테스트코드를 위해 private 을 제거하였습니다.
 
에러반환 여부만 테스트해도 되겠지만,
실무에서 예외처리를 하다보면 케이스별로 서로 다른 에러 객체를 throw 해줘야 하는 경우가 있는데,
그럼 public 한 에러 클래스를 구현하여 내부 로직과 테스트 코드에서 모두 사용하는 방식이 좋은 방법인가요?

nfreedom9 replied:
자바스크립트에서 저는 위와 같은 상황이 발생하면,
에러객체가 아니라 다른 객체 혹은 Symbole 타입을 throw 가 아닌 return 으로 처리하곤 하는데,
타입언어에서는 리턴타입을 자유롭게 다룰 수 없어 그런 방식이 불가능하네요. ㅎㅎ;
 
개인적으로는 `코드나 시스템상의 오류로 인한 에러` 와 `사용자의 입력에 의해 얼마든지 발생 가능한 에러` 를 구별하여 처리해야 한다고 생각하는데, 코틀린에서는 위와 같은 상황에서 에러처리를 어떤 방식으로 하는게 좋은 방법인가요?

laco-dev replied:
좋은 질문이네요. 🙏 

말씀주신 내용처럼 오류에도 여러가지 종류가 있다고 생각해요

`코드나 시스템상의 오류로 인한 에러` 
: 핵심적인 비즈니스 로직에 대한 오류로 명확한 예외로 표시를 해 주어야 한다고 생각해요. 만약 일반적인 예외 타입으로 구분이 어렵다면 명시적으로 예외 타입을 만들어서 넘겨주고는 합니다. 
이 과정에서 코틀린에서는 sealed class, sealed interface를 이용한 예외 케이스 정의 방법이 강력한 방법중 하나로 사용되고는 해요. 나중에 시간이 되실 때 관련된 키워드로 한번 검색 해 보세요
다만 미션을 진행하는 과정에서는 오류를 종류로 나눌정도로 복잡한 도메인을 가정하지 않아서 일반적인 예외 타입으로 비교하는 것으로 충분하다고 생각되어요
https://kotlinlang.org/docs/sealed-classes.html

`사용자의 입력에 의해 얼마든지 발생 가능한 에러`
: 이 경우 도메인에서 정의하고 테스트한 범위 외에 여러가지 예측이 불가능한 상황이 있을텐데요. 개발자가 캐치한 예외 뿐만 아니라 기상천외한 방식으로 터지기 때문에 저는 안전한 오류 처리를 지향하는 편입니다.
가령 자연수만 받는 객체를 만드는 경우 음수가 들어왔을때 예외를 던지기 보다는 음수인 경우 무조건 0으로 초기화를 한다거나 해서 만들고는 해요
다음 미션과정을 진행 하면서 자주 연습을 해 볼 기회가 있을거에요!

nfreedom9 commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,29 @@
+package step2
+
+object StringParser {
+    private val operatorSymbols = listOf<String>("+", "-", "*", "/") // 연산기호 문자열 리스트
```
연산기호에 해당하는 문자열을 이곳에서 따로 정의하고 있는데,
enum class 로 정의된 Operator 에서 실제로 사용되는 연산기호와 서로 다른 곳에서 정의하고 있는 점이 마음에 걸립니다.
 
가령, 지금은 아니지만 x 기호에 대하여 곱셈 연산으로 동작하게 하던지,
a, b, c 등 새로운 연산 기호에 대한 연산식을 정의하고 커스텀한 계산기 기능을 구현한다던지,
새로운 연산 기호와 연산식이 enum class 에 추가된다고 하면,
이곳에 정의된 operatorSymbols 에도 같은 연산기호를 추가해야 하는데,
Operator 에 있는 연산기호 리스트를 여기에서 알 수 있게 하는 방법은 없을까요?

nfreedom9 replied:
검색 후 수정하였습니다.

nfreedom9 committed:
operatorSymbols 변수를 Operator 클래스로 제공받도록 수정

nfreedom9 commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,23 @@
+package step2
+
+object StringParser {
+    private val operatorSymbols = Operator.symbols // 연산기호 문자열 리스트
```
https://github.com/next-step/kotlin-racingcar/pull/551#discussion_r874461530
검색 후 수정하였습니다.

nfreedom9 commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,23 @@
+    fun parseByOperatorSymbols(string: String): String = string
+        .split("")
+        .filter { it.isNotBlank() }
+        .joinToString("") { if (operatorSymbols.contains(it)) " $it " else it }
```
에러 throw 를 하지 않고, 문자열을 파씽하는 역할만 수행하도록 수정하였습니다.

nfreedom9 commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,23 @@
+        .joinToString("") { if (operatorSymbols.contains(it)) " $it " else it }
+
+    // 입력된 문자열로부터 연산기호를 기준으로 split 된 문자열 리스트 반환
+    fun listByOperatorSymbol(string: String): List<String> = parseByOperatorSymbols(string).split(" ")
```
기존에 문자열을 파씽하여 리스트로 변환하는 함수를 2 개의 함수로 분리하여 가독성과 로직을 이해하는데 개선하도록 하였습니다.
마찬가지로 이곳에서는 에러 throw 를 하지 않고, 주어진 역할만 수행하도록 하였습니다.

nfreedom9 commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,23 @@
+    } catch (e: NumberFormatException) {
+        println(e)
+        throw IllegalArgumentException(NumberFormatExceptionErrorMessage)
+    }
```
Calculator 에서 문자열을 숫자로 전환하며 예외처리 하던 것을
문자열을 다루는 StringParser 객체에 위임하여 처리하도록 수정하였습니다.

laco-dev commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+    SUBTRACT("-", { a: Double, b: Double -> a - b }),
+    MULTIPLY("*", { a: Double, b: Double -> a * b }),
+    DIVISION("/", { a: Double, b: Double ->
+        @Suppress("UNINITIALIZED_ENUM_COMPANION_WARNING")
+        if (b == 0.0) throw IllegalArgumentException(division0ErrorMessage)
+        a / b
```
코틀린의 require를 사용하면 보다 간결하게 예외를 표현할 수 있어요~
```kotlin
require(b != 0.0) { "message" }
```

laco-dev commented on src/main/kotlin/step2/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+package step2
+
+enum class Operator(val symbol: String, val operator: (Double, Double) -> Double) {
```
symbol을 찾는 방법을 안으로 내부적으로 숨겼기 때문에 더이상 외부에 노출 될 필요가 없어지겟죠!?
그러면 private 접근자로 바꿔 볼 수 있을 거에요!

nfreedom9 replied:
아.. Operator.values().map { it.symbol } 을 외부에서 실행하고 있었는데,
해당 값을 symbols 변수로 구현하였으니 private 으로 변경하는게 좋겠네용 :)

laco-dev commented on src/main/kotlin/step2/StringParser.kt:
```diff
@@ -0,0 +1,23 @@
+    // 입력된 문자열로부터 연산기호를 기준으로 split 된 문자열 리스트 반환
+    fun listByOperatorSymbol(string: String): List<String> = parseByOperatorSymbols(string).split(" ")
+
+    // 입력된 문자열로부터 Double 전환값을 반환. 에러 발생시 toDoubleErrorMessage 에러 throw
+    fun toDouble(string: String): Double = try {
+        string.toDouble()
+    } catch (e: NumberFormatException) {
+        println(e)
+        throw IllegalArgumentException(NumberFormatExceptionErrorMessage)
+    }
```
try, catch로 오류를 다시 전파하는 것도 좋은 방법인데요!
이 경우 toDoubleOrNull() 이라는 함수가 실패할 경우 null을 반환하도록 만들어서 코틀린에서는 아래와 같이 표현하기도 해요.
엘비스 연산자를 이용한 방법입니다!
```kotlin
fun toDouble(string: Double) = string.toDoubleOrNull() ?: throw IllegalArgumentException(...)
```

laco-dev replied:
다만 코틀린에서는 이미 표준 함수로 좋은 변환 방법을 제공하고 있어서 별도로 만들었을때 가지는 이점이 그렇게 크지는 않기도 하답니다.

nfreedom9 replied:
toDoubleOrNull 이라는 함수가 있었네요~ 
감사합니다.

laco-dev reviewed (approved):
열심히 반영 해 주셨네요. 👍 

질문주신 답변을 포함해서 코틀린에서 사용하는 방법 몇가지를 같이 남겨드렸어요.
3단계 미션을 먼저 PR 요청을 올려 주셔서 나중에 기회가 되실 때 커밋으로 진행중인 미션에서 같이 해결 해 주셔도 좋을 것  같아요.

자동차 경주 미션 해보면서 조금 더 연습 해보도록 하시죠 👍 
수고하셨습니다.
